keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+of such repair might be a `Database` custom resource. It follows the compiler pattern by producing a
+`StatefulSet` to deploy the database, but it extends that pattern to become an operator, because it
+also performs periodic backups of the database to cloud storage, and it can recover from those
+backups if all Pods in the StatefulSet fail. 
+
+### Open Questions
+
+#### Dependency management
+
+If truly core APIs are developed as custom resources, then there can be dependency issues when
+bootstrapping a Kubernetes cluster. If core components (e.g. the controller manager) needs to manage
+a custom resource, how do you ensure that the resource gets created before the controller-manager
+starts, when the controller-manager itself is in the lifecycle for custom resources. We need to untangle
+these circular dependencies. And ensure that people who write such controllers, write them in such a
+way that they can temporarily handle the resources they manage not existing.
+
+Another significant concern in dependency management is dependencies between custom resource definitions.
+As the ecosystem of plugins becomes richer, the chances that one plugin will depend on another increases.
+Currently there is no good way to express that in order for CustomResource `A` to work correctly, 
+CustomResource `B` needs to be installed as well.
+
+#### Version upgrades
+
+Another open question for CustomResources is version management. Currently there are no built-in practices
+for moving from one version of a CustomResource to another. To do such a migration, the tool itself would
+have to perform the schema migration by reading all of the data objects in the old version and rewriting
+them in the new version. To make this successful , at the very least we will need to provide the ability to
+mark a CustomResource as "read-only" to ensure that there isn't skew between the two representations.
+
+We may need to invest significantly more heavily in this space to make it easy for users to upgrade or 
+update their objects. For now, most users have simply chosen not to upgrade the version.
+
+### Risks and Mitigations
+
+#### DDOS or other attacks on storage
+
+Because custom resources share the same storage as the core API types, there is risk that either via 
+requests/second or storage space usage, custom resources will become deliberate or inadvertent denial
+of service attacks on a cluster. Currently there is a maximum size cap for custom resources, but it is
+a pretty limited defense against these sorts of concerns. This makes the installation of a custom
+resource a very privileged action for a cluster.
+
+#### Custom resource adminstration
+
+For many reasons, installing a new custom resource is a privileged action for a cluster adminstrator.
+However, this significantly increases the complexity of developing a Custom Resource. Ideally the
+resource itself would install itself into the cluster when it's controller is run as a Pod, but
```

Comment:
Given global discovery and cross namespace list/watch (as our apiserver operates), I don't agree with this statement.  The choice to add a schema'd object is privileged and affects the experience for the entire cluster.

Reply:
There is also the case where more than one controller is using the same CRD and they may have slightly different versions of it at the same CRD version. We can say people shouldn't do that but they already are doing it (e.g., updated descriptions or validations). This makes automation complicated.